ttg/ttg/serialization/data_descriptor.h

```diff
@@ -125,6 +120,92 @@ namespace ttg {
 
 #endif  // has MADNESS serialization
 
+#if defined(TTG_SERIALIZATION_SUPPORTS_BOOST)
+
+#include "ttg/serialization/backends/boost/archive.h"
+
+namespace ttg {
+
+  /// The default implementation for non-POD data types that are not directly copyable,
+  /// do not support MADNESS serialization, and support Boost serialization
+  template <typename T>
+  struct default_data_descriptor<
+      T, std::enable_if_t<(!std::is_trivially_copyable<T>::value && !detail::is_madness_buffer_serializable_v<T> &&
+                           detail::is_boost_buffer_serializable_v<T>) ||
+                          (!detail::is_madness_user_buffer_serializable_v<T> &&
+                           detail::is_boost_user_buffer_serializable_v<T>)>> {
+    static constexpr const bool serialize_size_is_const = false;
+
+    static uint64_t payload_size(const void *object) {
+      ttg::detail::boost_counting_oarchive oa;
+      oa << (*(T *)object);
+      return oa.streambuf().size();
```

Comment:
We need to serialize the object to find out it's serialized size? That means we serialize every object twice?

Reply:
Ahh yes, I didn't realize that the madness descriptor does the same. But why does the first serialize not perform stores? The object could be on the stack and not valid after the call so the archive has to make a copy.

We could allow the user to give us a hint at the size, e.g., by overloading `ttg::serialization::user_payload_size(T&)`. If that's not available we fall back to the double serialization approach.

Reply:
Yes, agreed, this is only the default implementation, when we can't do better.

Note that in `payload_size`, we use a `ttg::detail::boost_counting_oarchive`, and not a `ttg::detail::boost_oarchive`. `boost_counting_oarchive` should have an empty storage, and only do the counting.

Reply:
@devreal no way in general to compute data structure a priori in C++ except for trivial cases. Thus unfortunately must touch all data twice unless we want to make the 2-step mechanism only necessary for objects that exceed fixed size ... i.e. we could serialize into fixed-size buffer (user-configurable size, presumably as argument to ttg_initialize or smith) and if a given object exceeds its size we could use exceptions to signal that and invoke the 2-step mechanism.

I don't know enough about runtime performance impact of having exceptions in this path. But should be manageable. This will make serialization optimally fast for all but very large objects.